client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -0,0 +1,234 @@
+    }
+
+    @VisibleForTesting
+    static Long parseExpirationTime(String json) {
+        Long result = null;
+        if (json != null && !json.trim().equals("")) {
+
+            Matcher matcher = JWT_EXPIRATION_PATTERN.matcher(json);
+            if (matcher.find()) {
+               // Will look like this, if a match is found: "exp": 1569837434
+               String matchedString = matcher.group();
+
+               String[] expiryTimeFieldParts = matchedString.split(":");
+               if (expiryTimeFieldParts != null && expiryTimeFieldParts.length == 2) {
+                   result = Long.parseLong(expiryTimeFieldParts[1].trim());
+               }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Returns the delegation token. It returns existing delegation token if it is not close to expiry. If the token
+     * is close to expiry, it obtains a new delegation token and returns that one instead.
+     *
+     * @return String the delegation token JWT compact value
+     */
+    @Override
+    public String retrieveToken() {
+        if (this.delegationToken.get() == null) {
+            return this.refreshToken();
+        } else if (isTokenNearingExpiry()) {
+            log.info("Token is nearing expiry");
+            return refreshToken();
+        } else {
+            return delegationToken.get().getValue();
+        }
+    }
+
+    @Override
+    public boolean populateToken(@NonNull String token) {
+        if (token.trim().equals("")) {
+            return false;
+        } else {
+            this.delegationToken.set(new DelegationToken(token, extractExpirationTime(token)));
+            return true;
+        }
+    }
+
+    private boolean isTokenNearingExpiry() {
+        Long currentTokenExpirationTime = this.delegationToken.get().getExpiryTime();
+
+        // currentTokenExpirationTime can be null if the server returns a null delegation token
+        return currentTokenExpirationTime != null && isWithinRefreshThreshold(currentTokenExpirationTime);
+    }
+
+    private boolean isWithinRefreshThreshold(Long expirationTime) {
+        assert expirationTime != null;
+        return isWithinRefreshThreshold(Instant.now(), Instant.ofEpochSecond(expirationTime));
+    }
+
+    @VisibleForTesting
+    boolean isWithinRefreshThreshold(Instant currentInstant, Instant expiration) {
+        return currentInstant.plusSeconds(refreshThresholdInSeconds).getEpochSecond() >= expiration.getEpochSecond();
+    }
+
+    @VisibleForTesting
+    String refreshToken() {
+        long traceEnterId = LoggerHelpers.traceEnter(log, "refreshToken", this.scopeName, this.streamName);
+        CompletableFuture<Void> currentRefreshFuture = tokenRefreshFuture.get();
+        if (currentRefreshFuture == null) {
+            log.debug("Initiated token refresh for scope {} and stream {}", this.scopeName, this.streamName);
+            this.tokenRefreshFuture.set(this.recreateToken());
```

Comment:
This should be a compareAndSet incase some other thread raced this one.
This can be accomplished by creating a future and setting it, and then passing the future to recreate token so it can complete that.

Reply:
Sure, I've fixed it now.